src/main/java/org/jabref/logic/openoffice/oocsltext/MarkManager.java

```diff
@@ -0,0 +1,109 @@
+package org.jabref.logic.openoffice.oocsltext;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+
+import org.jabref.model.entry.BibEntry;
+
+import com.sun.star.container.XNameAccess;
+import com.sun.star.container.XNamed;
+import com.sun.star.lang.XMultiServiceFactory;
+import com.sun.star.text.XReferenceMarksSupplier;
+import com.sun.star.text.XTextDocument;
+import com.sun.star.uno.UnoRuntime;
+
+public class MarkManager {
+    private final HashMap<String, ReferenceMark> marksByName;
+    private final ArrayList<ReferenceMark> marksByID;
+    private final IdentityHashMap<ReferenceMark, Integer> idsByMark;
+    private final XTextDocument document;
+    private final XMultiServiceFactory factory;
+    private int lastUsedCitationNumber = 0;
+    private HashMap<String, Integer> citationKeyToNumber; // Add this line
+    private int highestCitationNumber = 0;
+
+    public MarkManager(XTextDocument document) throws Exception {
+        this.document = document;
+        this.factory = UnoRuntime.queryInterface(XMultiServiceFactory.class, document);
+        this.marksByName = new HashMap<>();
+        this.marksByID = new ArrayList<>();
+        this.idsByMark = new IdentityHashMap<>();
+        this.citationKeyToNumber = new HashMap<>();
+    }
+
+    public void readExistingMarks() throws Exception {
+        XReferenceMarksSupplier supplier = UnoRuntime.queryInterface(XReferenceMarksSupplier.class, document);
+        XNameAccess marks = supplier.getReferenceMarks();
+        for (String name : marks.getElementNames()) {
+            if (name.startsWith(CSLCitationOOAdapter.PREFIXES[0])) {
+                XNamed named = UnoRuntime.queryInterface(XNamed.class, marks.getByName(name));
+                ReferenceMark mark = new ReferenceMark(document, named, name);
+                addMark(mark);
+            }
+        }
+    }
+
+    public void addMark(ReferenceMark mark) {
+        marksByName.put(mark.getName(), mark);
+        idsByMark.put(mark, marksByID.size());
+        marksByID.add(mark);
+        updateCitationInfo(mark.getName());
+    }
+
+    private void updateCitationInfo(String name) {
+        String[] parts = name.split(" ");
+        if (parts.length >= 3) {
+            String citationKey = parts[1];
+            try {
+                int citationNumber = Integer.parseInt(parts[parts.length - 1]);
+                citationKeyToNumber.put(citationKey, citationNumber);
+                highestCitationNumber = Math.max(highestCitationNumber, citationNumber);
+            } catch (NumberFormatException e) {
+                // Ignore if we can't parse the number
```

Comment:
> On restarting JabRef, the numbering starts from 1 again for the same document.
Does logging these exceptions show a possible reason?

Reply:
Just tried, the exception is not triggered..